alibi_detect/od/tests/test__knn/test__knn_backend.py

```diff
@@ -0,0 +1,161 @@
+import pytest
+import torch
+
+from alibi_detect.od.pytorch.knn import KNNTorch
+from alibi_detect.utils.pytorch.kernels import GaussianRBF
+from alibi_detect.od.pytorch.ensemble import Ensembler, PValNormalizer, AverageAggregator
+from alibi_detect.od.base import NotFitException, ThresholdNotInferredException
+
+
+@pytest.fixture(scope='session')
+def ensembler(request):
+    return Ensembler(
+        normalizer=PValNormalizer(),
+        aggregator=AverageAggregator()
+    )
+
+
+def test_knn_torch_backend():
+    knn_torch = KNNTorch(k=5)
+    x = torch.randn((3, 10)) * torch.tensor([[1], [1], [100]])
+    x_ref = torch.randn((1024, 10))
+    knn_torch.fit(x_ref)
+    outputs = knn_torch.predict(x)
+    assert outputs.instance_score.shape == (3, )
+    assert outputs.is_outlier is None
+    assert outputs.p_value is None
+    scores = knn_torch.score(x)
+    assert torch.all(scores == outputs.instance_score)
+
+    knn_torch.infer_threshold(x_ref, 0.1)
+    outputs = knn_torch.predict(x)
+    assert torch.all(outputs.is_outlier == torch.tensor([False, False, True]))
+    assert torch.all(knn_torch(x) == torch.tensor([False, False, True]))
+
+
+def test_knn_torch_backend_ensemble(ensembler):
+    knn_torch = KNNTorch(k=[4, 5], ensembler=ensembler)
+    x_ref = torch.randn((1024, 10))
+    knn_torch.fit(x_ref)
+    x = torch.randn((3, 10)) * torch.tensor([[1], [1], [100]])
+    result = knn_torch.predict(x)
+    assert result.instance_score.shape == (3, )
+
+    knn_torch.infer_threshold(x_ref, 0.1)
+    outputs = knn_torch.predict(x)
+    assert torch.all(outputs.is_outlier == torch.tensor([False, False, True]))
+    assert torch.all(knn_torch(x) == torch.tensor([False, False, True]))
+
+
+def test_knn_torch_backend_ensemble_ts(tmp_path, ensembler):
+    knn_torch = KNNTorch(k=[4, 5], ensembler=ensembler)
```

Comment:
Would be good to add docstrings for tests i.e. assume `_ts` stands for `torchscript` but this might not be obvious for future readers.

Reply:
Added further comments

Reply:
I would just expand to have the test names be descriptive as there's no length cutoff - especially since you have used the fully qualified name in `test__knn.py` for `torchscript` already!